src/mesh/generic/Boundary.cpp

```diff
@@ -79,4 +93,64 @@ double Boundary::interpolate_prescribed_load(std::vector<std::pair<double, doubl
 
     return (1.0 - interpolation_factor) * load_0 + interpolation_factor * load_1;
 }
+
+void Boundary::generate_sinusoidal(json const& boundary,
+                                   std::string const& name,
+                                   double generate_time_step)
+{
+    using namespace ranges;
+
+    for (auto const& mandatory_field : {"Period", "Phase", "NumberOfCycles"})
+    {
+        if (!boundary.count(mandatory_field))
+        {
+            throw std::runtime_error("\"" + std::string(mandatory_field)
+                                     + "\" was not specified in \"BoundaryCondition\".");
+        }
+    }
+
+    std::vector<double> amplitude = boundary[name];
+    std::vector<double> period = boundary["Period"];
+    std::vector<double> number_of_cycles = boundary["NumberOfCycles"];
+    std::vector<double> phase = boundary["Phase"];
+
+    if (!((period.size() == phase.size()) && (phase.size() == number_of_cycles.size())))
+    {
+        throw std::runtime_error("The amplitude, period, number_of_cycles, time_step and phase "
+                                 "vectors must be of the same size.");
+    }
+
+    std::vector<double> new_times;
+    std::vector<double> new_loads;
+    new_times.push_back(0);
+
+    // Loop over the blocks
+    for (int i = 0; i < amplitude.size(); ++i)
+    {
+        double omega = 2.0 * M_PI / period[i];
+        auto number_of_steps = number_of_cycles[i] * period[i] / generate_time_step;
+
+        std::vector<double> time_block(number_of_steps);
+        std::iota(time_block.begin(), time_block.end(), 1);
+        auto scale_factor = number_of_cycles[i] * period[i] / time_block.back();
+        time_block = time_block | view::transform([scale_factor, new_times](double x) {
+                         return x * scale_factor + new_times.back();
+                     });
+
+        new_times.insert(new_times.end(), time_block.begin(), time_block.end());
+        new_loads = new_times | view::transform([i, amplitude, omega, phase](double x) {
+                        return amplitude[i] * std::sin(omega * x + phase[i]);
+                    });
+    }
+
+    json times = new_times;
+    json loads = new_loads;
+    // for (auto i : new_times) times.append(i);
+    // for (auto i : new_loads) loads.append(i);
+    *this = Boundary(times, loads);
+}
+void Boundary::generate_user_defined(json const& boundary, std::string const& name)
+{
+    *this = Boundary(boundary["Time"], boundary[name]);
```

Comment:
I haven't seen this pattern that much (at all).  What's the benefit to this, in comparison to an initialise function instead of this constructor call.  You can just call the initialise function with the parameters to do the checking etc.

Reply:
So, we can refactor what is in the constructor to just `allocate_time_load(json, json)` and the constructors just call this at the end, rather than using `*this`.  I think this is easier to understand.